src/vsr/grid_scrubber.zig

```diff
@@ -156,6 +159,71 @@ pub fn GridScrubberType(comptime Forest: type) type {
             scrubber.* = undefined;
         }
 
+        pub fn open(scrubber: *GridScrubber, random: std.rand.Random) void {
+            // Compute the tour origin exactly once.
+            if (scrubber.tour_tables_origin) |_| {
+                return;
+            }
+
+            // Each replica's scrub origin is chosen independently.
+            // This reduces the chance that the same block across multiple replicas can bitrot
+            // without being discovered and repaired by a scrubber.
+            //
+            // To accomplish this, try to select an origin uniformly across all blocks:
+            // - Bias towards levels with more tables.
+            // - Bias towards trees with larger Values (since their tables have more blocks).
+            // - (Though, for ease of implementation, the origin is always at the beginning of a
+            //   tree's level, never in the middle.)
+            assert(scrubber.tour == .init);
+
+            const tree_count = Forest.tree_infos.len;
+
+            var weights: [tree_count][constants.lsm_levels]u64 = undefined;
+            var weights_sum: u64 = 0;
+            for (0..constants.lsm_levels) |level| {
+                inline for (Forest.tree_infos, 0..) |tree_info, tree_index| {
+                    const tree_id = comptime Forest.tree_id_cast(tree_info.tree_id);
+                    const tree = scrubber.forest.tree_for_id_const(tree_id);
+                    const levels = &tree.manifest.levels;
+                    weights[tree_index][level] =
+                        levels[level].tables.len() * @sizeOf(tree_info.Tree.Table.Value);
+                    weights_sum += levels[level].tables.len();
+                }
+            }
+
+            scrubber.tour_tables_origin = origin: {
+                if (weights_sum == 0) {
+                    break :origin .{
+                        .level = 0,
+                        .tree_id = Forest.tree_infos[0].tree_id,
+                    };
+                }
+
+                var origin_offset = random.uintLessThan(u64, weights_sum);
+                for (0..constants.lsm_levels) |level| {
+                    inline for (Forest.tree_infos, 0..) |tree_info, tree_index| {
+                        if (origin_offset >= weights[tree_index][level]) {
+                            origin_offset -= weights[tree_index][level];
+                        } else {
+                            break :origin .{
+                                .level = @intCast(level),
+                                .tree_id = tree_info.tree_id,
+                            };
+                        }
+                    }
+                }
+                unreachable;
```

Comment:
Took me a second to convince myself that this is indeed unreacheable. I wonder if the code would end up simpler if we do it in one pass, without computing the array of weights. On a loop iteration, you have `weight_running_sum` and `weight_this_tree`, and toss a coin to decide whether the origin should be updated to point to the current tree.  So, on each iteration you pick a random tree from the set of trees considered so far:

```zig
            scrubber.tour_tables_origin = .{
                .level = 0,
                .tree_id = Forest.tree_infos[0].tree_id,
            };

            var weights_sum: u64 = 0;
            for (0..constants.lsm_levels) |level| {
                inline for (Forest.tree_infos) |tree_info| {
                    const tree_id = comptime Forest.tree_id_cast(tree_info.tree_id);
                    const tree = scrubber.forest.tree_for_id_const(tree_id);
                    const tree_level_weight =
                        tree.manifest.levels[level].tables.len() * @sizeOf(tree_info.Tree.Table.Value);
                    if (tree_level_weight > 0) {
                        weights_sum += tree_level_weight;
                        if (random.uintLessThan(u64, weights_sum) < tree_level_weight) {
                            scrubber.tour_tables_origin = .{
                                .level = level,
                                .tree_id = tree_info.tree_id,
                            };
                        }
                    }
                }
            }

```

Reply:
It is much simpler, but I'm having a hard time proving to myself that it is not biased towards the later levels/trees. :thinking:

Reply:
The loop invariant here is that we pick tree "uniformly" across the the trees considered so far. So the induction step is that we picked tree at random among the first n trees, and we need to decide whether to keep our choice, or to pick the n+1 tree. And we know the weight of this new tree, as well as the weights of trees so far.

I think this is called reservoir sampling: https://en.wikipedia.org/wiki/Reservoir_sampling

Reply:
Though, again, I am not sure what's simpler two loops which manage state, or one loop which tries to be smart.

Reply:
I like the reservoir sampling sampling version -- much easier to prove that it won't crash.